File: src/main/scala/org/clulab/timenorm/scfg/SynchronousGrammar.scala
package org.clulab.timenorm.scfg

import scala.collection.immutable.{IndexedSeq, Seq}

/**
 * A set of root symbols and synchronous rules that define a synchronous grammar.
 * 
 * Non-terminal symbols should be alphanumeric characters enclosed in square brackets, for example,
 * `[Period]`, `[TimeSpan]` or `[Int]`.
 * 
 * A single colon can be used to "subtype" a non-terminal, e.g. `[Int:4Digit]` or `[Unit:Singular]`.
 * In such cases, the symbol without the subtype information is called the "basic symbol", e.g.
 * `[Int]` and `[Unit]` are the basic symbols for the preceding two examples.
 * 
 * A symbol of the form `[Int:X-Y]`, where X and Y are integers, is treated as a numeric range
 * specification (inclusive of both endpoints). [[SynchronousParser]] has special handling of
 * such ranges.
 * 
 * A non-terminal whose basic symbol is `[Nil]` is called a "nil symbol". [[SynchronousParser]] has
 * special handling of such nil symbols. 
 *
 * @constructor Create a new grammar from a set of root symbols and a set of synchronous rules.
 * @param rootSymbols The symbols that are allowed to be the root of a parse.
 * @param rules The synchronous rules. 
 */
class SynchronousGrammar(val rootSymbols: Set[String], val rules: Seq[SynchronousGrammar.Rule]) {

  private val rulePrefixMap = PrefixMultiMap.empty[String, SynchronousGrammar.Rule]
  for (rule <- rules) {
    this.rulePrefixMap += (rule.sourceSeq, rule)
  }
  
  private val numberRegex = "^\\[Int:(\\d*)-(\\d*)\\]$".r
  private val numberRanges: Set[Range.Inclusive] = (
    for {
      rule <- rules
      numberRegex(begin, end) <- rule.symbol +: rule.sourceSeq
    } yield {
      begin.toInt to end.toInt
    }).toSet
  
  /**
   * Gets all non-terminal symbols whose range allows a particular number.
   * 
   * @param number The number whose possible non-terminal symbols are to be found.
   * @return Each non-terminal symbol whose range allows the number. 
   */
  def sourceSymbolsForNumber(number: Int): Set[String] = {
    val symbolsWithRanges =
      for (range <- this.numberRanges; if range.contains(number))
        yield "[Int:%d-%d]".format(range.start, range.end)
    symbolsWithRanges + "[Int]" 
  }

  /**
   * Gets all rules whose source side starts with a token sequence.
   * 
   * @param tokens The sequence of source tokens.
   * @return All rules whose source side starts with the given tokens.
   */
  def sourceSeqStartsWith(tokens: Seq[String]): Set[SynchronousGrammar.Rule] = {
    this.rulePrefixMap.getAllWithPrefix(tokens)
  }

  /**
   * Gets all rules whose source side starts with a token.
   * 
   * @param token The source token.
   * @return All rules whose source side starts with the given token.
   */
  def sourceSeqStartsWith(token: String): Set[SynchronousGrammar.Rule] = {
    this.rulePrefixMap.getAllWithPrefix(Seq(token))
  }
  
  /**
   * Gets all symbols used in the grammar.
   * 
   * This includes both terminal and non-terminal symbols
   * 
   * @return All symbols in the grammar.
   */
  def sourceSymbols(): Set[String] = {
    this.rules.flatMap(_.sourceSeq).toSet
  }
}

object SynchronousGrammar {

  /**
   * Determines whether a token is a terminal or non-terminal.
   * 
   * Currently, tokens must start with "[" and end with "]" to be a non-terminal.
   * 
   * @param token A token from a grammar.
   * @return True if the token is a terminal, false otherwise.
   */
  def isTerminal(token: String): Boolean = !token.matches("^\\[.*\\]$")

  /**
   * Strips any sub-type information from a non-terminal symbol.
   * 
   * For example, `[Int:4Digit]` would be converted to `[Int]`
   * 
   * @param token A non-terminal token from the grammar.
   * @return A non-terminal token without the sub-type information.
   */
  def basicSymbol(token: String): String = token.replaceAll(":[^\\]]*", "")

  /**
   * Determines whether a token is a number or not.
   * 
   * Currently, only tokens that are all digits are considered to be numbers.
   * 
   * @param token A token from a grammar.
   * @return True if the token is a number, false otherwise.
   */
  def isNumber(token: String): Boolean = token.matches("^\\d+$")
  
  /**
   * Determines whether a token is a nil non-terminal symbol or not.
   * 
   * Currently, only non-terminals whose basic symbol is "[Nil]" are considered to be nils.
   * 
   * @param token A token from a grammar.
   * @return True if the token is a nil non-terminal, false otherwise.
   */
  def isNil(token: String): Boolean = this.basicSymbol(token) == "[Nil]"
  
  /**
   * Parses a [[SynchronousGrammar]] from a string representation.
   * 
   * The first line defines the one or more root symbols and looks like:
   * <pre>
   * ROOTS [Period] [TimeSpan] ...
   * </pre>
   * The remaining lines follow the format of Joshua/Heiro
   * (http://joshua-decoder.org/4.0/file-formats.html) and look like:
   * <pre>
   * [Period] ||| [Period,1] and [Period,2] ||| Sum [Period,1] [Period,2] ||| 1.0
   * ...
   * </pre>
   * 
   * @param text The formatted grammar string.
   * @return A new [[SynchronousGrammar]].
   */
  def fromString(text: String): SynchronousGrammar = {
    val stripLabel: (String => String) = _.replaceAll("\\[(.*),.*\\]", "[$1]")
    // workaround instead of just `text.lines` because jdk11 also defines `String.lines`
    val lines = text.linesWithSeparators.map(_.stripLineEnd)
    val firstLine = lines.next
    if (!firstLine.startsWith("ROOTS")) {
      throw new IllegalArgumentException("First line must define root symbols, e.g. ROOTS [XXX] [YYY]")
    }
    val rootSymbols = firstLine.split("\\s+").tail.toSet
    val nonCommentLines = lines.filterNot(_.startsWith("//"))
    val rules = for (line <- nonCommentLines) yield line.trim.split("\\s*[|][|][|]\\s*") match {
      case Array(symbol, sourceSeqString, targetSeqString, scoreString) => {
        val sourceSeqItems = sourceSeqString.split("\\s+").toIndexedSeq
        val targetSeqItems = targetSeqString.split("\\s+").toIndexedSeq
        val sourceNonTerminals = sourceSeqItems.filterNot(this.isTerminal)
        val targetNonTerminals = targetSeqItems.filterNot(this.isTerminal)
        val alignment = for ((token, targetIndex) <- targetNonTerminals.zipWithIndex) yield {
          if (sourceNonTerminals.count(_ == token) != 1) {
            val message = "Expected exactly 1 non-terminal matching \"%s\" in \"%s\""
            throw new IllegalArgumentException(message.format(token, sourceSeqString))
          }
          targetIndex -> sourceNonTerminals.indexOf(token)
        }
        Some(Rule(symbol, sourceSeqItems.map(stripLabel), targetSeqItems.map(stripLabel), alignment.toMap))
      }
      case Array("") => None
      case _ => throw new IllegalArgumentException("\"" + line + "\"")
    }
    new SynchronousGrammar(rootSymbols, rules.flatten.toList)
  }

  /**
   * A synchronous grammar rule.
   * 
   * @constructor Creates a synchronous grammar rule.
   * @param symbol A non-terminal symbol
   * @param sourceSeq The sequence of source tokens
   * @param targetSeq The sequence of target tokens
   * @param nonTerminalAlignment A mapping from source non-terminal indexes to the corresponding
   *        target non-terminal indexes. For example, <code>Map(1->2, 2->1)</code> would indicate
   *        that the first source non-terminal is the second target non-terminal and vice versa. 
   */
  case class Rule(symbol: String, sourceSeq: IndexedSeq[String], targetSeq: IndexedSeq[String], nonTerminalAlignment: Map[Int, Int]) {
    
    /**
     * Strips any sub-type information from this rule's symbol.
     * 
     * See [[SynchronousGrammar.basicSymbol]].
     */
    val basicSymbol = SynchronousGrammar.basicSymbol(this.symbol)
    
    /**
     * Determines whether this rule's symbol is a nil non-terminal symbol or not.
     *
     * See [[SynchronousGrammar.isNil]].
     */
    val isNil = SynchronousGrammar.isNil(this.symbol)
  }
}

private[timenorm] class PrefixMultiMap[K, V] {

  var suffixes = Map.empty[K, PrefixMultiMap[K, V]]
  var values = Set.empty[V]

  def +=(key: Seq[K], value: V): Unit = {
    if (key.isEmpty) {
      this.values += value
    } else {
      val head = key.head
      if (!this.suffixes.contains(head)) {
        this.suffixes += head -> new PrefixMultiMap[K, V]
      }
      this.suffixes(head) += (key.tail, value)
    }
  }

  def get(key: Seq[K]): Set[V] = {
    this.getMap(key) match {
      case None => Set.empty
      case Some(map) => map.values
    }
  }

  def getAll: Set[V] = {
    this.values ++ this.suffixes.values.flatMap(_.getAll)
  }

  def getAllWithPrefix(key: Seq[K]): Set[V] = {
    this.getMap(key) match {
      case None => Set.empty
      case Some(map) => map.getAll
    }
  }

  private def getMap(key: Seq[K]): Option[PrefixMultiMap[K, V]] = {
    if (key.isEmpty) {
      Some(this)
    } else {
      this.suffixes.get(key.head).flatMap(_.getMap(key.tail))
    }
  }
}

private[timenorm] object PrefixMultiMap {
  def empty[K, V] = new PrefixMultiMap[K, V]
}

File: src/main/python/synchronous_grammar.py
import re


class SynchronousGrammar(object):

    def __init__(self, root_symbols, rules):
        self.root_symbols = root_symbols
        self.rules = rules
        self._rule_prefix_map = _PrefixMultiMap()
        self._number_ranges = []
        for rule in self.rules:
            self._rule_prefix_map.add(rule.sources, rule)
            for symbol in [rule.symbol] + rule.sources:
                match_number_ranges = re.match(r"^\[Int:(\d*)-(\d*)\]$", symbol)
                if match_number_ranges is not None:
                    begin = int(match_number_ranges.group(1))
                    end = int(match_number_ranges.group(2))
                    number_range = list(range(begin, end+1))
                    self._number_ranges.append(number_range)

    def source_symbols_for_number(self, number):
        symbols_with_ranges = []
        for number_range in self._number_ranges:
            if number in number_range:
                symbols_with_ranges.append("[Int:%d-%d]" % (number_range[0], number_range[-1]))
        symbols_with_ranges.append("[Int]")
        return symbols_with_ranges

    def source_seq_startswith(self, tokens):
        if not isinstance(tokens, list):
            tokens = [tokens]
        return self._rule_prefix_map.get_all_with_prefix(tokens)

    def source_symbols(self):
        source_symbol_list = []
        for rule in self.rules:
            source_symbol_list.extend(rule.sources)
        return source_symbol_list


class _PrefixMultiMap(object):

    def __init__(self):
        self.suffixes = {}
        self.values = []

    def add(self, keys, value):
        if not bool(keys):
            self.values.append(value)
        else:
            head = keys[0]
            if head not in self.suffixes:
                self.suffixes[head] = _PrefixMultiMap()
            self.suffixes[head].add(keys[1:], value)

    def _get_map(self, keys):
        if not bool(keys):
            return self
        else:
            head = keys[0]
            if head in self.suffixes:
                return self.suffixes[head]._get_map(keys[1:])
            else:
                return None

    def get(self, keys):
        prefix_multi_map = self._get_map(keys)
        if prefix_multi_map is None:
            return []
        else:
            return prefix_multi_map.values

    def get_all(self):
        all_values = []
        all_values.extend(self.values)
        for suffix_values in self.suffixes.values():
            all_values.extend(suffix_values.get_all())
        return all_values

    def get_all_with_prefix(self, keys):
        prefix_multi_map = self._get_map(keys)
        if prefix_multi_map is None:
            return []
        else:
            return prefix_multi_map.get_all()


class Rule(object):

    def __init__(self, symbol, sources, targets, non_terminal_alignment):
        self.symbol = symbol
        self.sources = sources
        self.targets = targets
        self.non_terminal_alignment = non_terminal_alignment
        self.basic_symbol = basic_symbol(self.symbol)
        self.is_nil = is_nil(self.symbol)


def is_terminal(token):
    return not re.search(r"^\[.*\]$", token)


def is_number(token):
    return re.search(r"^\d+$", token)


def basic_symbol(token):
    return re.sub(r":[^\]]*", "", token)


def is_nil(token):
    return basic_symbol(token) == "[Nil]"


def strip_label(label):
    return re.sub(r'\[([^\[\]]*),[^\[\]]*\]', r"[\1]", label)


def load_grammar(grammar_path):
    with open(grammar_path) as grammar_file:
        lines = grammar_file.read().splitlines()
        first_line = lines[0]
        assert first_line.startswith("ROOTS"), "First line must define root symbols, e.g. ROOTS [XXX] [YYY]"
        root_symbols = re.split(r"\s+", first_line)[1:]
        rules = []
        for line in lines[1:]:
            if not (line.startswith("//") or line == ""):
                symbol, sources_string, targets_string, score_string = re.split(r"\s*[|][|][|]\s*", line)
                sources = re.split("\\s+", sources_string)
                targets = re.split("\\s+", targets_string)
                source_non_terminals = list(filter(lambda t: not is_terminal(t), sources))
                target_non_terminals = list(filter(lambda t: not is_terminal(t), targets))
                alignment = {}
                for target_idx, token in enumerate(target_non_terminals):
                    if source_non_terminals.count(token) != 1:
                        raise ValueError("Expected exactly 1 non-terminal matching \"%s\" in %s" % (token, sources))
                    alignment[target_idx] = source_non_terminals.index(token)
                sources = list(map(strip_label, sources))
                targets = list(map(strip_label, targets))
                rule = Rule(symbol, sources, targets, alignment)
                rules.append(rule)
        return SynchronousGrammar(root_symbols, rules)
